harper-core/src/linting/for_noun.rs

```diff
@@ -0,0 +1,77 @@
+use crate::{
+    Token,
+    patterns::{AnyCapitalization, NominalPhrase, OwnedPatternExt, Pattern, SequencePattern},
+};
+
+use super::{Lint, LintKind, PatternLinter, Suggestion};
+
+pub struct ForNoun {
+    pattern: Box<dyn Pattern>,
+}
+
+impl Default for ForNoun {
+    fn default() -> Self {
+        let pattern = SequencePattern::aco("fro")
+            .then_whitespace()
+            .then(NominalPhrase.or(Box::new(AnyCapitalization::of("sure"))));
+
+        Self {
+            pattern: Box::new(pattern),
+        }
+    }
+}
+
+impl PatternLinter for ForNoun {
+    fn pattern(&self) -> &dyn Pattern {
+        self.pattern.as_ref()
+    }
+
+    fn match_to_lint(&self, matched_tokens: &[Token], source: &[char]) -> Option<Lint> {
+        let span = matched_tokens.first()?.span;
+        let problem_chars = span.get_content(source);
+
+        Some(Lint {
+            span,
+            lint_kind: LintKind::WordChoice,
+            suggestions: vec![Suggestion::replace_with_match_case_str(
+                "for",
+                problem_chars,
+            )],
+            message: "`For` is more common in this context.".to_owned(),
+            priority: 31,
+        })
+    }
+
+    fn description(&self) -> &'static str {
+        "Corrects the archaic or mistaken `fro` to `for` when followed by a noun."
```

Comment:
```suggestion
        "Corrects the archaic or mistaken `fro` to `for`, or `from` when followed by a noun."
```

It could be a typo about `fro` used instead of `from`

> I received a letter fro Sarah
> I received a letter from Sarah

Reply:
Also, while `fro` is archaic, it's irrelevant since it's not an archaic form of `for`. It's an archaic form of `from` and is only used these days in `to and fro`.

I wonder if some logic for rare/archaic words could have a flag in the dictionary to allow them to get a warning and still undergo spellcheck. We'd want to find other instances similar to `fro` for that to be worthwhile. I know that back in the day spellcheck dictionaries purposefully left words out that were correct but rare and likely to cause confusion.

Reply:
Would you suggest we change the description, or remove this rule as well as the dictionary entry for `fro`?

Reply:
> Would you suggest we change the description, or remove this rule as well as the dictionary entry for `fro`?

I would probably test if the spellcheck handles it. I've noticed it often doesn't make what I consider the 'obvious' suggestions for many mistakes. I haven't checked this one.

If the spellcheck suggests 'for' and 'from' we can probably do without the rule, but if it fails to suggest those then it's probably worth keeping the rule with a changed description.

At some point we probably want to add some code to 'score' suggestions in other ways than just pure edit distance to help get the best 3 into the suggestions.

Reply:
That's been around for a while now. See `harper-core/src/spell/mod.rs` `fn score_suggestion`. I've tweaked it a bit, but I'm not sure how to achieve the desired behavior.